### PR TITLE
Implement our own config explorer

### LINKED ISSUE #15653: Improving config resolution performance
❤️ First of all, I would to say a big **THANK YOU** to all the maintainers of this great project! I use Prettier on a daily basis, and I have to say that greatly improved my DX. ❤️

After seeing the [interest on Twitter](https://x.com/Vjeux/status/1722733472522142022?s=20) (sorry on X...) for working on a faster alternative to Prettier written in Rust. I was wondering how much could we potentially improve the current Prettier implementation without having to do a full rewrite in Rust.

## Observation

For this purpose, I decided to use the [mui/material-ui](https://github.com/mui/material-ui) project to benchmark Prettier. This project contains ~42K files and roughly 8500 files files that are currently formatted by Prettier. On my machine, it takes 84s to format this project with the [latest](https://github.com/prettier/prettier/commit/3478719cc09e434f2621720389c38ea17a8a61ff) version of the main branch (median over 5 invocations, caching disabled, OSX, Node v20.6.1).

After further investigation, it appears that a sizable amount of time is spent in [`getOptionsForFile`](https://github.com/prettier/prettier/blob/3478719cc09e434f2621720389c38ea17a8a61ff/src/cli/format.js#L336-L339). To validate this observation, I commented the `getOptionsForFile` call and inlined the project configuration. Formatting the entire material-ui project now takes 46s, which is a **45% improvement**.

The fact that it takes 45% of the processing time to resolve file formatting options is highly suspicious. There are certainly some performance improvement opportunities here.

## Digging deeper

As far as I understand, the `getOptionsForFile` essentially resolves and loads the prettier and editor config files in a given directory. Looking at this code it appears to me that Prettier could get a great performance boost from caching the configuration data between getOptionsForFile invocations.

To put this in perspective, I logged all the file system invocations when formatting the material-ui project, the option resolution logic issues: 

* `fs.promises.access`: 827K invocations
* `fs.existsSync`: 100K invocations
* `fs.readFile`: 55K invocations
* `fs.promises.readFile`: 13K invocations

The option resolution logic is delegated to 2 libraries: lilconfig and editorconfig. Both libraries do essentially the same thing, traversing up the file system and looking for specific configuration files. Currently, the configuration lookup isn’t cached between invocations resulting in repeated file system calls.

The main issue here is that it is currently difficult for Prettier to cache the config resolution result between invocations:

* Prettier recently moved cosmiconfig from lilconfig. Unlike the previous config resolution library, lilconfig doesn’t have [cache](https://github.com/antonk52/lilconfig#difference-to-cosmiconfig) support and it appears to be a non-goal for this project. 
* Prettier currently pins the editorconfig dependency to version 0.15.3 but caching support was only introduced in version  [1.0.0](https://github.com/editorconfig/editorconfig-core-js/releases/tag/v1.0.0). As far as I understand, bumping the version isn’t an option because Prettier can’t bundle WASM modules. 

Misc observations:
* [`findProjectRoot`](https://github.com/prettier/prettier/blob/8158a8948a9c94aded9c989660e5c5337ff46ca0/src/config/find-project-root.js) function also recursively traverse the file system upward looking for `.git` or `.hg` directories. Caching the result of its invocation would certainly yield good performance improvement. It would also be ideal to replace the synchronous file system lookup with an async one to avoid blocking the main thread, but I don’t expect major performance improvements as all the files are currently processed serially.
* I think #15363 should be revived. As mentioned in the PR, invoking `.search` on the file itself results in unnecessary file system lookups. It should be possible to optimize this without introducing a breaking change by checking first if the filename references a directory or not. 

## Moving forward

I would be happy to make adequate changes, but I would love to get the maintainers’ feedback first before I invest time into this.

1. I don’t understand the reason for migrating from cosmiconfig to lilconfig ([PR](https://github.com/prettier/prettier/pull/15232)). Is it an option to move back to cosmiconfig to leverage its [caching capabilities](https://github.com/cosmiconfig/cosmiconfig#caching)? If not, I would then open an issue on lilconfig to determine if the maintainer is interested in adding caching support.
2. I don’t see a straightforward way to add support for WASM to the build in order to bump the editorconfig dependency. We can either ask if the maintainer is willing to make the parser swappable to work around this. An alternative approach would be re-implement .editorconfig parsing logic in Prettier. The processing logic looks small enough not to add too much tech debt.  Any thoughts on this?

### PR DESCRIPTION
**Pull request by fisker**
## Description

Ref #15653

## Checklist

- [ ] I’ve added tests to confirm my change works.
- [ ] (If changing the API or CLI) I’ve documented the changes I’ve made (in the `docs/` directory).
- [ ] (If the change is user-facing) I’ve added my changes to `changelog_unreleased/*/XXXX.md` file following `changelog_unreleased/TEMPLATE.md`.
- [ ] I’ve read the [contributing guidelines](https://github.com/prettier/prettier/blob/main/CONTRIBUTING.md).

**✨[Try the playground for this PR](https://prettier.io/playground-redirect)✨**

### CODE REVIEW

**Comment by fisker** [comment]
Accidentally committed https://github.com/prettier/prettier/commit/b5313a5a5abdf9418bef646a51854002fd66094d https://github.com/prettier/prettier/commit/ba258537e309914f88bec0aea99319a6bd589647 to "main" branch 😢

File: src/config/prettier-config-explorer/config-searcher.js [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+      (await isPackageJsonFileWithPrettierConfig(file))
+    ) {
+      return file;
+    }
+  }
+}
+
+/** @type {Map} */ // @ts-expect-error -- intentionally not add the `get` method
+const noopMap = { has: () => false, set() {} };
+class Searcher {
```
Comment (sosukesuzuki) [inline src/config/prettier-config-explorer/config-searcher.js:37]: I prefer align this name with as used one:

```suggestion
class ConfigSearcher {
```

File: src/config/prettier-config-explorer/config-searcher.js [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+  #cache;
```
Comment (sosukesuzuki) [inline src/config/prettier-config-explorer/config-searcher.js:38]: Is this a just boolean value? If true I prefer naming as `shouldCache`. To clarify whether this is an object that has cached values or a just boolean:

```suggestion
  #shouldCache;
```

File: src/config/prettier-config-explorer/config-searcher.js [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+  #stopDirectory;
+  #store;
```
Comment (sosukesuzuki) [inline src/config/prettier-config-explorer/config-searcher.js:40]: I would also like a name that makes it easy to understand what is being stored. What do you think `searchedFilesStore` ?

File: src/config/prettier-config-explorer/common.js [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+      return cache.get(fileOrDirectory);
+    }
+
+    const promise = function_(fileOrDirectory);
+    cache.set(fileOrDirectory, promise);
+    const result = await promise;
+    cache.set(fileOrDirectory, result);
+    return result;
+  };
+}
```
Comment (sosukesuzuki) [inline src/config/prettier-config-explorer/common.js:44]: This function is only used in `config-searcher.js`, right? I prefer to move from `common.js` to `config-searcher.js` and name the function something like `createCachedSearchFunction`. The name createCachedFunction is too generic.

File: src/config/prettier-config-explorer/loaders.js [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+  const { prettier } = await readJson(file);
+  return prettier;
+}
+
+async function loadYaml(file) {
+  const content = await readFile(file);
+  try {
+    return jsYaml.load(content);
+  } catch (/** @type {any} */ error) {
+    error.message = `YAML Error in ${file}:\n${error.message}`;
```
Comment (sosukesuzuki) [inline src/config/prettier-config-explorer/loaders.js:33]: ~Can we use Error cause?~
Sorry we still support Node.js 14

**Review by sosukesuzuki — COMMENTED** [review]
Thank you for your great job!! Overall it looks good, but I did comment on some points that were difficult to understand while I reading the code. Please check.

**Comment by phocks** [comment]
This appears to have broken VS Code formatting Svelte files with `prettier-plugin-svelte` somehow :/ https://github.com/prettier/prettier-vscode/issues/3253#issuecomment-1968173844

**Pull request merged**